inference-engine/src/mkldnn_plugin/nodes/mkldnn_shuffle_channels_node.cpp

```diff
@@ -0,0 +1,712 @@
+                        vpermq(ymm_dst, ymm_dst, 0x08);
+                    uni_vpackuswb(vmm_dst, vmm_dst, vmm_dst);
+                    if (isa != cpu::sse42)
+                        vmovq(op, xmm_dst);
+                    else
+                        movd(op, xmm_dst);
+                }
+                break;
+            default:
+                assert(!"unknown dst_dt");
+        }
+    }
+
+    inline void store_scalar(const Xbyak::Address &op, Xmm xmm_dst, memory::data_type dst_dt) {
+        if (dst_dt != memory::f32) {
+            uni_vcvtps2dq(xmm_dst, xmm_dst);
+        }
+
+        switch (dst_dt) {
+            case memory::f32:
+            case memory::s32:
+                movss(op, xmm_dst);
+                break;
+            case memory::s8:
+                uni_vpackssdw(xmm_dst, xmm_dst, xmm_dst);
+                uni_vpacksswb(xmm_dst, xmm_dst, xmm_dst);
+                movq(reg_tmp_64, xmm_dst);
+                mov(op, reg_tmp_8);
+                break;
+            case memory::u8:
+                uni_vpackusdw(xmm_dst, xmm_dst, xmm_dst);
+                uni_vpackuswb(xmm_dst, xmm_dst, xmm_dst);
+                movq(reg_tmp_64, xmm_dst);
+                mov(op, reg_tmp_8);
+                break;
+            default:
+                assert(!"unknown dst_dt");
+        }
+    }
+};
+
+MKLDNNShuffleChannelsNode::MKLDNNShuffleChannelsNode(const InferenceEngine::CNNLayerPtr& layer, const mkldnn::engine& eng, MKLDNNWeightsSharing::Ptr &cache)
+        : MKLDNNNode(layer, eng, cache) {}
+
+void MKLDNNShuffleChannelsNode::getSupportedDescriptors() {
+    if (!descs.empty())
+        return;
+
+    if (getParentEdges().empty() || getChildEdges().empty())
+        THROW_IE_EXCEPTION << "ShuffleChannels layer with name " << getName() << " gets incorrect number of input/output edges!";
+
+    if (getParentEdgeAt(0)->getDims().ndims() != getChildEdgeAt(0)->getDims().ndims()) {
+        THROW_IE_EXCEPTION << "ShuffleChannels layer with name " << getName() << " gets incorrect number of input/output dimensions!";
+    }
+
+    auto *layer = getCnnLayer().get();
+    dst_dims = layer->outData[0]->getTensorDesc().getDims();
+    dims_size = dst_dims.size();
+
+    axis = layer->GetParamAsInt("axis", 1);
+    if (axis < 0)
+        axis += dims_size;
+
+    if (axis < 0 || axis >= static_cast<int>(dims_size))
+        THROW_IE_EXCEPTION << "ShuffleChannels layer with name " << getName() << " gets incorrect input parameters dimensions and axis number!";
+
+    group = layer->GetParamAsUInt("group", 1);
+    if (group == 0 || dst_dims[axis] % group)
+        THROW_IE_EXCEPTION << "ShuffleChannels layer with name " << getName() <<
+                              " gets incorrect group parameter! Group parameter must evenly divide the channel dimension!";
+
+    shuffle_size = dst_dims[axis];
+    group_size = shuffle_size / group;
+
+    if (dims_size <= 5) {
+        if (dims_size == 5) {
```

Comment:
In fact I don't like this approach because it is not cheap to extend (for example for 6D).
Looks like ShuffleChannels with arbitrary axis and tensor problem can be simplified to the following description: input/output tensors should be expressed as 3D [O, A , I], where A - axis dimension, O - multiplication of outer dims, I - multiplication of inner dims (example SC with input [5, 3, 10, 4, 6] and axis 2 will be expressed as [15, 10, 24] with axis = 1). Given that your implementation should only support 3D tensors with axis always = 1 and this will work for any real input rank. O - dimension can be fully used for parallelization. Jit kernel will implement shuffling for full A dimension.
As for different layouts support:
1) In case layout = nhwc, ndhwc, etc you just need to adjust axis parameter (example: if you have ShCh layer with axis = 1, but layout is nhwc you can reinterpret this as ShCh layer with logical layout nchw but axis = 3; in that case 3D task representation will look like: [n * h * w, c, 1]);
2) Case with blocked layout (nChw8c, nCdhw8c etc) and shuffling by channel cannot be so easily simplified.  From the top of my head I would advise to use 4D representation: [O, A_num_of_blocks , I, A_block_size]. For simplifying I would advise to support blocked layouts only for case block_size % group_size  == 0 or group_size % block_size ==0;

Reply:
Great suggestion! Arbitrary dimensions are expressed as 3D to be supported by JIT for planar layout. In blocked layout, JIT shuffling is fully supported by the instructions of gather.